### DISCUSSION TITLE
Disable Turborepo cache when using Vercel's deploy hooks for SSG

### DISCUSSION THREAD

**Discussion by saifbechan** [vercel/community#433] [Help]
We just switched over to Turborepo. A problem we are facing now is that when we use the webhook for one of our projects we still get the old data. We use SSG and Turborepo caches this and we have found no way to purge this only for the webhook.

We could disable Turborepo's cache, or make a code change and deploy that, but we would rather use a webhook that would deploy a freshly build app.

**Comment by jaronheard** [comment]
Same issue! Seems like the Turborepo cache should be ignored based on the [`?buildCache=false` flag described in the docs.](https://vercel.com/docs/concepts/git/deploy-hooks#build-cache).

Even having a different build hook param for the Turborepo cache could be fine, but it seems right now that you just need to do one of the workarounds described above.

**Reply by rshackleton** [reply]
A separate flag would be useful, throwing the entire build cache out means all npm dependencies need to be reinstalled. Really all we need is to force build whichever package(s) with the external dependency.

Perhaps an extension of turborepo to support declaring external dependencies would also be a long-term solution.

**Comment by saifbechan** [comment]
Any update on this issue?

**Comment by RizBizKits** [comment]
👋 @saifbechan Interesting! Hold tight, I'll speak with our Engineering team to confirm this.

**Reply by saifbechan** [reply]
Hello @RizBizKits thanks for your reply. Is there any update from the engineering team. Still we are facing the problem that we cannot just use the "redeploy" option in Vercel to have the content changed from our CMS. Every time when we need content changed we have to do a code update.

**Reply by samuelhorn** [reply]
Also interested in this since I have the exact same problem!

**Reply by RizBizKits** [reply]
Hi @samuelhorn! Looking back, I can see a PR was merged that was related to this issue. Can you provide more details of what you're experiencing as well as the version of `turbo` that you're using?

**Reply by samuelhorn** [reply]
Hi @RizBizKits!

Tried upgrading to the latest version but still the same issue. 

So basically I get a cache hit every time the project is deployed from the CMS through a webhook (when nothing in the code has changed). The only way I found to get around the problem is using --force in the build command on Vercel to bypass all caching. On the webhook call, I have the parameter `?buildCache=false` defined, which doesn't seem to help.

**Reply by destroytoday** [reply]
Curious about this as well, as I’m experiencing this now when using a deploy hook with Sanity. I added `?buildCache=false` from the docs, but still no luck. Do we also need to specify something to bypass the ignored build step?

**Reply by paulkuhle** [reply]
I believe the `?buildCache=false` param in the deploy hook only affects Vercel's build cache (your `node_modules` basically), but not the Turborepo cache (the actual build output). I'd say you do want to keep the Vercel build cache so that you don't install your packages again.

Turborepo allows you to include environment variables in the hash function. For example, say you have the following build step in your `turbo.json`:

```
"build": {
    "dependsOn": ["^build", "$NEXT_PUBLIC_VERCEL_ENV"],
    "outputs": ["dist/**", ".next/**"]
},
```

Without the `$NEXT_PUBLIC_VERCEL_ENV"` dependency, Turborepo would cache your build between the different environments, which could lead to issues for example if your preview deployments are supposed to use a staging API. If Turborepo caches the build, either your production or your preview deployments would point to the wrong API, depending on which environment is in the build cache. This is a useful API for this use case as well, at least theoretically.

I think the main blocker here with regards to the headless CMS deployments is that Turborepo doesn't know where the build was triggered from because Vercel doesn't expose any environment variable that tells you whether the build comes from the Git integration, a deploy hook or a manual trigger. Even then, you'd have to make some changes to the Turborepo hashing algorithm because of course the variable wouldn't change between invocations.

If the goal is to not make any changes to Turborepo, Vercel would have to expose an environment variable that includes your deploy hook name and a unique invocation ID, although I'm not sure if there are any other use cases for something like this.

Another idea would be to make the turbo.json file dynamic (similar to tsconfig, eslint, etc.), although I think currently the JSON file is parsed directly in Go.
```
module.exports = {
    pipeline: {
        build: {
            outputs: ["dist/**", ".next/**"],
            cache: process.env.VERCEL_DEPLOY_HOOK === 'my-cms-hook' ? false : true,
    },
}
```

Or you could write a custom build script that checks if that environment variable is present and executes `turbo run build` with the `--force` flag, overrides your `turbo.json` file with one that has the `"cache": false` argument or something along those lines. It wouldn't be a super neat integration but a good workaround if we had that environment variable.

Something like this maybe:
```
cd ../..
if [ "$VERCEL_DEPLOY_HOOK" == "my-hook" ]
    then
        mv turbo.json turbo.temp.json
        cp turbo.no-build-cache.json turbo.json
fi
turbo run build ...
mv turbo.temp.json turbo.json
```

For the time being, I have deactivated the build cache in the config file:
```
"build": {
    "dependsOn": ["^build", "$NEXT_PUBLIC_VERCEL_ENV"],
    "outputs": ["dist/**", ".next/**"],
    "cache": false
},
```

**Reply by Pieparker** [reply]
We have recently launched the `VERCEL_FORCE_NO_BUILD_CACHE` env var, and configuring this with a value of `1` will prevent the build cache from being used. Does this help your use case?

**Reply by bobbyflowstate** [reply]
Hey @Pieparker, I'm running into this issue as well. I'd like to use a webhook (deploy hook) on vercel to force build every day so that new data is incorporated into our static pages. How can I use `VERCEL_FORCE_NO_BUILD_CACHE` env var for this use case? 

When I create a deploy hook in vercel it only asks for a name and a branch:
<img width="932" alt="Screen Shot 2022-12-16 at 6 05 46 PM" src="https://user-images.githubusercontent.com/2475027/208203046-6e333659-4e16-45b8-9d6c-762349a53755.png">

How would I dynamically set the env variable `VERCEL_FORCE_NO_BUILD_CACHE=1` from the above, vercel deploy hook?

I guess I could set it to 1 for my main branch always...

**Reply by bobbyflowstate** [reply]
I just tried that and it still doesn't work...

Is it because I have the following command in "Ignore Build Step":  `npx turbo-ignore`?

<img width="577" alt="Screen Shot 2022-12-16 at 6 12 00 PM" src="https://user-images.githubusercontent.com/2475027/208203497-f4b317fe-4b9d-4a2f-8544-cfe0ef95dddf.png">

**Reply by amyegan** [reply]
Hi Ashok, sorry I missed this last week! I know you already figured out that `turbo-ignore` was skipping the build because it couldn't detect CMS content changes, but wanted to share a link to our other discussion for anyone else who comes across this. We are talking about other options in https://github.com/vercel/community/discussions/1170
